File: 2017/Day05.js
module.exports = function()
{
    const fs = require('fs');
    const readline = require('readline');

    const readInput = readline.createInterface({
        input: fs.createReadStream('Data/Day05.data')
    });

    let input1 = [];
    let input2 = [];

    readInput
    .on('line', (line) => { 
        processLine(line);
    })
    .on('close', () => {
        jump(input1, false)
        jump(input2, true);

        process.exit(0);
    });

    function processLine(line)
    {
        let n = +line;
        
        if (isNaN(n))
            throw 'invalid data';
        input1.push(n);
        input2.push(n);
    }

    function jump(input, part2)
    {
        let current = 0;
        let steps = 0;

        let count = input.length;
        
        while(current < count && current >= 0)
        {
            steps++;
            let offset = input[current];   
            if (part2 && offset >= 3)
                input[current]--;         
            else
                input[current]++;
            current += offset;
        }
        console.log((part2 ? 'Part 2' : 'Part 1') + ' executed in ' + steps + ' steps');
    }
}

File: 2017/C/Day05.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>

static int _input_1[1024];
static int _input_2[1024];
static int _count = 0;

static long solve(int* input, int part)
{
    long steps  = 0;
    int current = 0;

    while (current >= 0 && current < _count)
    {
        steps++;

        int offset = input[current];

        if (part == 2 && offset >= 3)
            input[current]--;
        else
            input[current]++;

        current += offset;
    }

    return steps;
}

static void loadData()
{
    FILE* file = fopen("Data/Day05.data", "r");

    if (file == NULL)
    {
        printf("File not found!\n");
		return;
    }
        
    static char buffer[32];

	int count = 0;
    while (! feof(file))
    {
        char* ptr = fgets(buffer, 32, file);
        int value = atoi(ptr);
		
		_input_2[count] =
        _input_1[count] = value;
		
		count++;
    }
	
	_count = count;
    fclose(file);
}

void day5(void)
{
	double ms = CLOCKS_PER_SEC / 1000;
	
	long start = clock();
	loadData();
	long end = clock();
	
	printf("Data loaded in %lf ms\n\n", (end-start) / ms);
	
	long result_1 = solve(_input_1, 1);
	long result_2 = solve(_input_2, 2);
	
	printf("%li steps for part 1 ( expected 318883 )\n", result_1);
    printf("%li steps for part 2 ( expected 23948711 )\n", result_2);
}
